File: TheWatch-iOS/ViewModels/HomeViewModel.swift
import Foundation
import MapKit
import Combine

@Observable
final class HomeViewModel {
    var userLocation: CLLocationCoordinate2D = CLLocationCoordinate2D(latitude: 37.7749, longitude: -122.4194)
    var nearbyResponders: [Responder] = []
    var communityAlerts: [CommunityAlert] = []
    var activeAlert: Alert?
    var isLoading = false
    var errorMessage: String?
    var sosCountdown = 0
    var isSOS = false
    var showNavigationDrawer = false
    var isPhraseDetectionActive = false
    var lastPhraseMatch: PhraseMatchResult?

    private let alertService: any AlertServiceProtocol
    private let volunteerService: any VolunteerServiceProtocol
    private let apiClient = WatchApiClient.shared
    private let hubConnection = WatchHubConnection.shared
    private let phraseDetectionService = PhraseDetectionService.shared
    private var sosTimer: Timer?
    private var cancellables = Set<AnyCancellable>()
    private var responderPollingTask: Task<Void, Never>?

    init(
        alertService: any AlertServiceProtocol,
        volunteerService: any VolunteerServiceProtocol
    ) {
        self.alertService = alertService
        self.volunteerService = volunteerService

        // Subscribe to phrase match results — route to SOS pipeline
        phraseDetectionService.matchResultPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] result in
                self?.handlePhraseMatch(result)
            }
            .store(in: &cancellables)
    }

    // MARK: - Phrase Detection

    func startPhraseDetection() {
        Task {
            let authorized = await phraseDetectionService.requestAuthorization()
            guard authorized else {
                errorMessage = "Speech recognition permission required for phrase detection"
                return
            }
            phraseDetectionService.startListening()
            isPhraseDetectionActive = true
        }
    }

    func stopPhraseDetection() {
        phraseDetectionService.stopListening()
        isPhraseDetectionActive = false
    }

    /// Route phrase match results to the appropriate action:
    /// - Duress → silent SOS (no visible alert on screen)
    /// - ClearWord → cancel active alert
    /// - Custom → standard SOS with countdown
    private func handlePhraseMatch(_ result: PhraseMatchResult) {
        lastPhraseMatch = result
        guard let phrase = result.matchedPhrase else { return }

        print("[HomeViewModel] Phrase match: type=\(phrase.type), text=\"\(phrase.phraseText)\", confidence=\(result.confidence)")

        switch phrase.type {
        case .duress:
            // Silent SOS — skip countdown, no visible UI change
            triggerSilentSOS(description: "Duress phrase detected: silent SOS")

        case .clearWord:
            // Cancel active alert — user confirmed safe
            if activeAlert != nil {
                cancelSOS()
                activeAlert = nil
            }

        case .custom:
            // Standard SOS trigger
            triggerSOSAlert()
        }
    }

    private func triggerSilentSOS(description: String) {
        Task {
            do {
                let alert = try await alertService.createAlert(
                    userId: "user-001",
                    severity: .critical,
                    type: .duress,
                    latitude: userLocation.latitude,
                    longitude: userLocation.longitude,
                    description: description
                )
                self.activeAlert = alert
                // No haptic feedback, no visible countdown — silent
            } catch {
                errorMessage = "Failed to send silent SOS"
            }
        }
    }

    func loadNearbyData() async {
        isLoading = true
        errorMessage = nil

        async let responders = volunteerService.getNearbyResponders(
            latitude: userLocation.latitude,
            longitude: userLocation.longitude,
            radiusMeters: 5000
        )

        async let alerts = alertService.getNearbyAlerts(
            latitude: userLocation.latitude,
            longitude: userLocation.longitude,
            radiusMeters: 5000
        )

        do {
            self.nearbyResponders = try await responders
            self.communityAlerts = try await alerts
        } catch {
            errorMessage = "Failed to load nearby data"
        }

        isLoading = false
    }

    func initiateSOS() {
        sosCountdown = 3
        isSOS = true

        sosTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.sosCountdown -= 1
            if self?.sosCountdown == 0 {
                self?.sosTimer?.invalidate()
                self?.triggerSOSAlert()
            }
        }

        // Haptic feedback
        let impact = UIImpactFeedbackGenerator(style: .heavy)
        impact.impactOccurred()
    }

    func cancelSOS() {
        sosTimer?.invalidate()
        sosCountdown = 0
        isSOS = false

        let impact = UIImpactFeedbackGenerator(style: .medium)
        impact.impactOccurred()
    }

    private func triggerSOSAlert() {
        Task {
            do {
                let alert = try await alertService.createAlert(
                    userId: "user-001",
                    severity: .critical,
                    type: .sos,
                    latitude: userLocation.latitude,
                    longitude: userLocation.longitude,
                    description: "SOS Button Activated"
                )
                self.activeAlert = alert
                self.isSOS = false

                // Start polling for responder positions (real-time via SignalR
                // when hub is live; polling as fallback in mock mode)
                startResponderPolling(requestId: alert.id)

                // Additional haptic pulses
                for _ in 0..<3 {
                    try await Task.sleep(nanoseconds: 200_000_000)
                    let feedback = UINotificationFeedbackGenerator()
                    feedback.notificationOccurred(.Success)
                }
            } catch {
                errorMessage = "Failed to send SOS alert"
            }
        }
    }

    func triggerSOS() async {
        triggerSOSAlert()
    }

    func toggleNavigationDrawer() {
        showNavigationDrawer.toggle()
    }

    /// Start polling for responder locations during an active alert.
    /// This is a fallback when SignalR is in mock mode; the real hub
    /// provides instant ResponderLocationUpdated events.
    func startResponderPolling(requestId: String) {
        responderPollingTask?.cancel()
        responderPollingTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { break }
                do {
                    let situation = try await self.apiClient.getSituation(requestId: requestId)
                    let responders = (situation.responders ?? []).map { ack in
                        Responder(
                            id: ack.responderId ?? "",
                            name: ack.responderName ?? "",
                            role: ResponderRole(rawValue: ack.responderRole ?? "Volunteer") ?? .volunteer,
                            latitude: ack.latitude ?? 0,
                            longitude: ack.longitude ?? 0,
                            distance: ack.distanceMeters ?? 0,
                            skills: [],
                            isVerified: true,
                            responseTime: nil,
                            status: .onCall,
                            availability: .available,
                            hasVehicle: ack.hasVehicle ?? false
                        )
                    }
                    await MainActor.run {
                        self.nearbyResponders = responders
                    }
                } catch {
                    // Polling failure is non-fatal; try again next cycle
                }
                try? await Task.sleep(nanoseconds: 5_000_000_000) // 5 seconds
            }
        }
    }

    func stopResponderPolling() {
        responderPollingTask?.cancel()
        responderPollingTask = nil
    }

    deinit {
        sosTimer?.invalidate()
        responderPollingTask?.cancel()
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/ui/screens/home/HomeViewModel.kt
package com.thewatch.app.ui.screens.home

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.thewatch.app.data.model.Responder
import com.thewatch.app.data.repository.AlertRepository
import com.thewatch.app.data.repository.PhraseDetectionRepository
import com.thewatch.app.data.repository.PhraseMatchResult
import com.thewatch.app.data.repository.PhraseType
import com.thewatch.app.data.signalr.WatchHubConnection
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

sealed class AlertUiState {
    object Idle : AlertUiState()
    object Loading : AlertUiState()
    object AlertActive : AlertUiState()
    object AlertCancelled : AlertUiState()
    data class Error(val message: String) : AlertUiState()
}

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val alertRepository: AlertRepository,
    private val phraseDetectionRepository: PhraseDetectionRepository,
    private val hubConnection: WatchHubConnection
) : ViewModel() {
    private val _alertUiState = MutableStateFlow<AlertUiState>(AlertUiState.Idle)
    val alertUiState: StateFlow<AlertUiState> = _alertUiState.asStateFlow()

    private val _isAlertActive = MutableStateFlow(false)
    val isAlertActive: StateFlow<Boolean> = _isAlertActive.asStateFlow()

    private val _unreadNotifications = MutableStateFlow(2)
    val unreadNotifications: StateFlow<Int> = _unreadNotifications.asStateFlow()

    /** Nearby responders fetched from API / SignalR. */
    private val _nearbyResponders = MutableStateFlow<List<Responder>>(emptyList())
    val nearbyResponders: StateFlow<List<Responder>> = _nearbyResponders.asStateFlow()

    /** Active response request ID, if any. */
    private val _activeRequestId = MutableStateFlow<String?>(null)
    val activeRequestId: StateFlow<String?> = _activeRequestId.asStateFlow()

    /** Whether phrase detection is currently listening. */
    val isPhraseDetectionActive: StateFlow<Boolean> = phraseDetectionRepository.isListening

    /** Last phrase match result (for debug/status UI). */
    private val _lastPhraseMatch = MutableStateFlow<PhraseMatchResult?>(null)
    val lastPhraseMatch: StateFlow<PhraseMatchResult?> = _lastPhraseMatch.asStateFlow()

    init {
        // Collect phrase match results and route to SOS pipeline
        viewModelScope.launch {
            phraseDetectionRepository.matchResults.collect { result ->
                handlePhraseMatch(result)
            }
        }
    }

    /**
     * Route phrase match results to the appropriate action:
     * - Duress → silent SOS (no visible alert on screen)
     * - ClearWord → cancel active alert
     * - Custom → standard SOS with countdown
     */
    private fun handlePhraseMatch(result: PhraseMatchResult) {
        _lastPhraseMatch.value = result
        val phrase = result.matchedPhrase ?: return

        Log.i("HomeViewModel", "Phrase match: type=${phrase.type}, text=\"${phrase.phraseText}\", confidence=${result.confidence}")

        when (phrase.type) {
            PhraseType.DURESS -> {
                // Silent SOS — no visible countdown, no screen alert
                activateAlert(alertType = "Duress", description = "Duress phrase detected: silent SOS", silent = true)
            }
            PhraseType.CLEAR_WORD -> {
                // Cancel active alert — user confirmed safe
                if (_isAlertActive.value) {
                    cancelAlert()
                }
            }
            PhraseType.CUSTOM -> {
                // Standard SOS trigger
                activateAlert(alertType = "Emergency", description = "Emergency phrase detected")
            }
        }
    }

    fun activateAlert(
        alertType: String = "Emergency",
        description: String = "Emergency assistance needed",
        silent: Boolean = false
    ) {
        viewModelScope.launch {
            _alertUiState.value = if (silent) AlertUiState.AlertActive else AlertUiState.Loading
            try {
                // Alert.kt has default params for all fields
                val alert = com.thewatch.app.data.model.Alert(
                    userId = "current_user", // TODO: inject actual user ID from auth
                    latitude = 40.7128,
                    longitude = -74.0060,
                    description = description,
                    severity = alertType
                )
                val result = alertRepository.activateAlert(alert)
                result.onSuccess { activated ->
                    _activeRequestId.value = activated.id
                    _alertUiState.value = AlertUiState.AlertActive
                    _isAlertActive.value = true
                    // Start loading nearby responders for map overlays
                    loadNearbyResponders(activated.latitude, activated.longitude)
                }
                result.onFailure { e ->
                    _alertUiState.value = AlertUiState.Error(e.message ?: "Unknown error")
                }
            } catch (e: Exception) {
                _alertUiState.value = AlertUiState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun cancelAlert() {
        viewModelScope.launch {
            _alertUiState.value = AlertUiState.Loading
            try {
                val requestId = _activeRequestId.value ?: ""
                alertRepository.cancelAlert(requestId)
                _alertUiState.value = AlertUiState.AlertCancelled
                _isAlertActive.value = false
                _activeRequestId.value = null
                _nearbyResponders.value = emptyList()
            } catch (e: Exception) {
                _alertUiState.value = AlertUiState.Error(e.message ?: "Unknown error")
            }
        }
    }

    /**
     * Load nearby responders from the API. During an active alert, this also
     * starts a polling flow that updates every 5 seconds (SignalR provides
     * instant updates when the real hub connection is active).
     */
    fun loadNearbyResponders(latitude: Double = 40.7128, longitude: Double = -74.0060) {
        viewModelScope.launch {
            try {
                alertRepository.getNearbyResponders(latitude, longitude, 5000.0)
                    .collect { responders ->
                        _nearbyResponders.value = responders
                    }
            } catch (e: Exception) {
                Log.w("HomeViewModel", "Failed to load nearby responders: ${e.message}")
            }
        }
    }

    fun startPhraseDetection() {
        phraseDetectionRepository.startListening()
    }

    fun stopPhraseDetection() {
        phraseDetectionRepository.stopListening()
    }

    fun markNotificationsAsRead() {
        _unreadNotifications.value = 0
    }
}
